File: app/src/main/java/com/agotakiss/androidtest/presentation/detail/DetailsActivity.java
package com.agotakiss.androidtest.presentation.detail;

import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.method.ScrollingMovementMethod;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.agotakiss.androidtest.R;
import com.agotakiss.androidtest.domain.models.Genre;
import com.agotakiss.androidtest.domain.models.Movie;
import com.agotakiss.androidtest.domain.repository.MovieRepository;
import com.agotakiss.androidtest.injector.Injector;
import com.agotakiss.androidtest.presentation.BaseActivity;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.BiFunction;
import io.reactivex.schedulers.Schedulers;

import static com.agotakiss.androidtest.presentation.home.MovieAdapter.IMAGE_BASE_URL;
import static com.agotakiss.androidtest.presentation.home.MovieAdapter.MOVIE;

public class DetailsActivity extends BaseActivity {

    private MovieRepository movieRepository = Injector.getMovieRepository();

    private int similarMoviesPage = 1;
    private int totalPages;
    private String movieId;
    private Movie movie;
    List<Movie> similarMovieList = new ArrayList<>();
    private RecyclerView recyclerView;
    private SimilarMovieAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_details);
        movie = (Movie) getIntent().getSerializableExtra(MOVIE);
        movieId = Integer.toString(movie.getId());
        initUI();
        loadSimilarMovies();
        initializeSimilarMovieList();
    }

    private void loadSimilarMovies() {
//        MovieDbManager.getInstance().loadSimilarMovies(movieId, similarMoviesPage, new Callback<LoadMoviesResponse>() {
//            @Override
//            public void onResponse(Call<LoadMoviesResponse> call, Response<LoadMoviesResponse> response) {
//               onSimilarMoviesLoaded(response.body().getMovies(), response.body().getTotalPages());
//            }
//
//            @Override
//            public void onFailure(Call<LoadMoviesResponse> call, Throwable t) {
//                Toast.makeText(DetailsActivity.this, "Error: " + t.toString(), Toast.LENGTH_SHORT).show();
//            }
//        });
        movieRepository.getSimilarMovies(movieId, similarMoviesPage)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(movieList -> {
                    onSimilarMoviesLoaded(movieList, Integer.MAX_VALUE);
                }, throwable -> {
                    logE(throwable);
//                    Toast.makeText(DetailsActivity.this, "Error: " + throwable.toString(), Toast.LENGTH_SHORT).show();
                });
    }

    private void initUI() {
        ImageView posterDetailsImageView = findViewById(R.id.poster_detailed);
        TextView movieTitleDetailsTextView = findViewById(R.id.movie_title_detailed);
        TextView genresDetailedTextView = findViewById(R.id.genres_detailed);
        TextView ratingDetailedTextView = findViewById(R.id.rating_detailed);
        TextView releaseDateDetailedTextView = findViewById(R.id.release_date_detailed);
        TextView descriptionDetailedTextView = findViewById(R.id.description_detailed);
        ImageView backdropDetailedImageView = findViewById(R.id.backdrop_detailed);

        Picasso.get().load(IMAGE_BASE_URL + movie.getPosterPath()).into(posterDetailsImageView);
        Picasso.get().load(IMAGE_BASE_URL + movie.getBackdropPath()).into(backdropDetailedImageView);
        backdropDetailedImageView.setAlpha(0.2f);

        movieTitleDetailsTextView.setText(movie.getTitle());
        genresDetailedTextView.setText(genresToString());
        ratingDetailedTextView.setText(Float.toString(movie.getAverageVote()));
        releaseDateDetailedTextView.setText(movie.getReleaseDateText().substring(0, 4));
        descriptionDetailedTextView.setText(movie.getOverview());
        descriptionDetailedTextView.setMovementMethod(new ScrollingMovementMethod());
    }

    private void onSimilarMoviesLoaded(List<Movie> newMovies, int pages) {
        similarMovieList.addAll(newMovies);
        adapter.notifyItemRangeInserted(similarMovieList.size() - newMovies.size(), newMovies.size());
        totalPages = pages;
    }

    public void initializeSimilarMovieList() {
        recyclerView = findViewById(R.id.similar_recycler_view);
        LinearLayoutManager layoutManager = new LinearLayoutManager(DetailsActivity.this, LinearLayoutManager.HORIZONTAL, false);
        recyclerView.setLayoutManager(layoutManager);
        adapter = new SimilarMovieAdapter(similarMovieList, DetailsActivity.this);
        recyclerView.setAdapter(adapter);
        adapter.setOnEndReachedListener(position -> {
            similarMoviesPage++;
            if (similarMoviesPage < totalPages) {
                loadSimilarMovies();
            }
        });
    }

    public String genresToString() {
//        List<Integer> movieGenreIdList = movie.getGenreIdList();
//        if (movie.getGenreIdList() != null && genresMap != null) {
//            StringBuilder stringBuilder = new StringBuilder();
//            for (int i = 0; i < movie.getGenreIdList().size() - 1; i++) {
//                stringBuilder.append(genresMap.get(movieGenreIdList.get(i)));
//                stringBuilder.append(", ");
//            }
//            stringBuilder.append(genresMap.get(movieGenreIdList.get(movieGenreIdList.size() - 1)));
//            return stringBuilder.toString();
//        } else return "";
        if(movie.getGenres() != null && !movie.getGenres().isEmpty()){
            return Observable.fromIterable(movie.getGenres())
                    .map(Genre::getName)
                    .reduce((s, s2) -> s + ", " + s2).blockingGet();
        } else{
            return "";
        }
    }

}

File: app/src/main/java/com/agotakiss/androidtest/presentation/detail/DetailsActivity.kt
package com.agotakiss.androidtest.presentation.detail

import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.ActivityOptionsCompat
import android.support.v4.view.ViewCompat
import android.support.v7.widget.LinearLayoutManager
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import com.agotakiss.androidtest.R
import com.agotakiss.androidtest.domain.models.Cast
import com.agotakiss.androidtest.domain.models.Movie
import com.agotakiss.androidtest.presentation.BaseActivity
import com.agotakiss.androidtest.presentation.actor.ActorDetailsActivity
import com.agotakiss.androidtest.presentation.main.MainActivity
import com.agotakiss.androidtest.presentation.main.MovieAdapter.Companion.IMAGE_BASE_URL
import com.agotakiss.androidtest.presentation.main.MovieAdapter.Companion.MOVIE
import com.agotakiss.androidtest.presentation.main.OnEndReachedListener
import com.squareup.picasso.Picasso
import io.reactivex.Observable
import kotlinx.android.synthetic.main.activity_details.*
import java.util.*
import javax.inject.Inject

class DetailsActivity : BaseActivity(), DetailsView {

    companion object {
        const val ACTOR_TRANSITION_NAME = "actorTransition"
    }

    val applicationComponent by lazy { movieApplication.applicationComponent.plus(DetailsModule(this)) }

    @Inject
    lateinit var presenter: DetailsPresenter

    private lateinit var movie: Movie
    internal var similarMovieList: MutableList<Movie> = ArrayList()
    internal var actorsList: MutableList<Cast> = ArrayList()
    private lateinit var adapter: SimilarMovieAdapter
    private lateinit var actorAdapter: ActorAdapter
    private var hasStartedAnotherScreen = false

    private var sharedImageView: View? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_details)
        applicationComponent.inject(this)
        movie = intent.getSerializableExtra(MOVIE) as Movie
        initUI()
        initializeSimilarMovieList()
        initializeActorsList()
        presenter.onViewReady(this, movie)
        description_detailed.setOnClickListener { expandCollapsedByMaxLines(description_detailed) }
    }

    private fun initUI() {
        setSupportActionBar(toolbar)
        setTitle(movie.title)

        Picasso.get().load(IMAGE_BASE_URL + movie.posterPath).into(poster_detailed)
        Picasso.get().load(IMAGE_BASE_URL + movie.backdropPath).into(backdrop_detailed)
        Picasso.get().load(IMAGE_BASE_URL + movie.backdropPath).into(collapsing_image)
        backdrop_detailed.alpha = 0.2f
        movie_title_detailed.text = movie.title
        genres_detailed.text = genresToString()
        rating_detailed.text = java.lang.Float.toString(movie.averageVote)
        release_date_detailed.text = movie.releaseDateText!!.substring(0, 4)
        description_detailed.text = movie.overview
    }

    fun initializeSimilarMovieList() {
        val layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        similar_recycler_view.layoutManager = layoutManager
        adapter = SimilarMovieAdapter(similarMovieList, object : OnEndReachedListener {
            override fun onEndReached(position: Int) {
                presenter.onSimilarMovieScrollEndReached()
            }
        }) { movie, view ->
            ViewCompat.setTransitionName(poster_detailed, "random")
            sharedImageView = view
            hasStartedAnotherScreen = true
            val similarMovie = movie
            val intent = Intent(this, DetailsActivity::class.java)
            intent.putExtra(MOVIE, similarMovie)
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation(this, view,
                MainActivity.POSTER_TRANSITION_NAME)
            startActivity(intent, options.toBundle())
        }
        similar_recycler_view.adapter = adapter
    }

    override fun onResume() {
        super.onResume()

        if (hasStartedAnotherScreen) {
            hasStartedAnotherScreen = false
            ViewCompat.setTransitionName(sharedImageView, "unused")
            ViewCompat.setTransitionName(poster_detailed, MainActivity.POSTER_TRANSITION_NAME)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            android.R.id.home -> {
                supportFinishAfterTransition()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        supportFinishAfterTransition()
    }

    override fun showSimilarMovies(newMovies: List<Movie>) {
        similarMovieList.addAll(newMovies)
        adapter.notifyItemRangeInserted(similarMovieList.size - newMovies.size, newMovies.size)
    }

    fun initializeActorsList() {
        val layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        actors_recycler_view.layoutManager = layoutManager
        actorAdapter = ActorAdapter(actorsList, object : OnEndReachedListener {
            override fun onEndReached(position: Int) {
                presenter.onSimilarMovieScrollEndReached()
            }
        }) { actorId, view ->
            ViewCompat.setTransitionName(poster_detailed, "random")

            val intent = Intent(this, ActorDetailsActivity::class.java)
            intent.putExtra(ActorAdapter.ACTOR_ID, actorId)
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation(this, view,
                ACTOR_TRANSITION_NAME)
            startActivity(intent, options.toBundle())
        }
        actors_recycler_view.adapter = actorAdapter
    }

    override fun showActors(newActors: List<Cast>) {
        actorsList.addAll(newActors)
        actorAdapter.notifyItemRangeInserted(actorsList.size - newActors.size, newActors.size)
    }

    override fun showError(t: Throwable) {
        Toast.makeText(this, "Error loading the movies.", Toast.LENGTH_LONG).show()
        logE(t)
    }

    fun genresToString(): String {
        return if (movie.genres != null && !movie.genres!!.isEmpty()) {
            Observable.fromIterable(movie.genres!!)
                .map<String> { it.name }
                .reduce { s, s2 -> "$s, $s2" }.blockingGet()
        } else {
            ""
        }
    }

    @SuppressLint("Range")
    fun expandCollapsedByMaxLines(text: TextView) {
        val height = text.measuredHeight
        text.height = height
        text.maxLines = Integer.MAX_VALUE //expand fully
        text.measure(View.MeasureSpec.makeMeasureSpec(text.measuredWidth, View.MeasureSpec.EXACTLY),
            View.MeasureSpec.makeMeasureSpec(ViewGroup.LayoutParams.WRAP_CONTENT, View.MeasureSpec.UNSPECIFIED))
        val newHeight = text.measuredHeight
        val animation = ObjectAnimator.ofInt(text, "height", height, newHeight)
        animation.setDuration(250).start()
    }
}
